Resubmission of templating PR (#1747)

ehiggs opened the pull request:
This is a manual rebase of the older #1747 PR which was basically unrebasable due to the time it's been dormant.. This implements templating for Cargo and is basically the work of Greg Chapple (@gchp).

I'd love for this feature to move forward since it's really tedious to create all the directories (e.g. `src/bin`) and copy paste docopt examples which I then edit. Then implement the `main` program to delegate to the subcommands in `src/bin`, etc.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @alexcrichton (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

killercup commented:
Nice! I would really love to have this. Copying the same README.md, LICENSE, travis, gitignore, editorconfig, etc. files into every repo can't be the way it's supposed to go.

gchp commented:
Thanks for re-submitting this. I never got the time to come back to it unfortunately, but I think its very useful and would love to see it finished off and merged. Thanks for taking it on @ehiggs!

gchp commented:
The biggest thing not addressed by this from the earlier PR was the use of Handlebars instead of Mustache. At the time the existing Handlebars implementations were a little unstable, but I imagine it's better now so should be easier to work in.

ehiggs commented:
@gchp agreed. So the work is as follows:

- [ ] Migrate to handlebars.
- [ ] Remove caching mechanism. If users want a local version, they will have to manage the download and copying into `~/.cargo/template` manually.
- [ ] Tests
- [ ] Documentation

Future work: `cargo template` suite of subcommands for managing templates.

alexcrichton commented on Cargo.lock:
```diff
@@ -441,16 +460,16 @@ version = "0.7.14"
 [[package]]
 name = "openssl-sys"
-version = "0.7.14"
+version = "0.7.17"
```
It looks like this may have run `cargo update`, but could this commit just add the dependencies it needs to?

alexcrichton commented:
Thanks for the revival @ehiggs! Excited to see progress on this

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -412,74 +423,123 @@ fn mk(config: &Config, opts: &MkOptions) -> CargoResult<()> {
+            };
+            let template_dir = templates_dir.join(repo_name);
+
+            match Repository::clone(template, &*template_dir) {
```
Cargo typically issues a status message for operations that may take a long time, so perhaps this could indicate that a repository is being cloned?

I think this may also wish to use the helpers in `git/utils.rs` to ensure this handles pieces such as authentication, retries, etc.

ehiggs replied:
I added the note. But I wasnt sure what you meant by `git/utils.rs`. I didn't see anything [here](https://github.com/alexcrichton/git2-rs/blob/master/src/util.rs) to do what you suggested.

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -412,74 +423,123 @@ fn mk(config: &Config, opts: &MkOptions) -> CargoResult<()> {
+                          template.starts_with("git@") => {
+            let path = PathBuf::from(template);
+
+            let repo_name = match path.components().last().unwrap() {
```
This `unwrap()` should be `chain_error` w/ a propagation of the error

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -548,6 +608,95 @@ fn global_config(config: &Config) -> CargoResult<CargoNewConfig> {
+}
+
+/// Create a new "bin" project
+fn create_bin_template(path: &PathBuf) -> CargoResult<()> {
```
It seems like these should perhaps be "built in" rather than literally created on the filesystem?

alexcrichton commented on src/doc/guide.md:
```diff
@@ -408,6 +411,65 @@ will not fail your overall build. Please see the [Travis CI Rust
 documentation](https://docs.travis-ci.com/user/languages/rust/) for more
 information.
 
+# Templates
```
This documentation doesn't seem to mention `--template name` along with `~/.cargo/templates`, but maybe that won't be used that much in practice?

ehiggs replied:
It says:

```
# use the mytemplate template which is located in .cargo/templates/mytemplate   
$ cargo new myproj --template mytemplate   
```

ehiggs commented:
I've done some work on this and rebased. I changed the way remote repos work by cloning them into a tempdir and using them as a template. The tempdir is then deleted when it goes out of scope. I do something similar with the default bin/lib templates. These are dumped to a tempdir and then used as a template.

I tried to use `cargo::sources::git::utils::GitRemote` and friends but when I try to clone the remote repository using `checkout` it seems to only copy the `.git` directory. I guess this is used to finely figure out which revisions are available or maybe I misunderstood how it's supposed to work since I thought there might be a nice single function to call to get all the work done.

In trying to get the `git::utils` working, I noticed that it only supports remote git repositories with Urls. This means remote repositories defined using ssh strings (e.g. `git@github.com:ehiggs/rust-cmdline-skeleton.git`) won't work.

I also made a [skeleton repository](https://github.com/ehiggs/rust-cmdline-skeleton) to test this with. It used docopt and has a LICENSE file and everything to get started writing command line programs that use subcommands. It's basically @BurntSushi's [xsv](https://github.com/BurntSushi/xsv) stripped down to... a skeleton (I keep calling it a skeleton since that's what it's called in [`adduser`](http://linuxcommand.org/man_pages/adduser8.html).

ehiggs commented:
I've added some commits that address docs and fix the tests.

ehiggs commented:
Travis failed due to a network issue. Appveyor failed on "override_cargo_home" but I don't have access to a Windows box to try and reproduce it.

alexcrichton commented:
Thanks for the update @ehiggs! Some thoughts of mine:

* The management of git templates still feels like we need to tweak the UI here. For example once it's cloned the first time it'll never get updated until it's removed? I'd expect that each time you create a project it'd update the git repo currently.

* Storing custom templates in `~/.cargo/templates` seems like it may be brittle wrt namespace collisions? It's also kinda weird to store user data in a dotfile location. I'd personally prefer if `~/.cargo/templates` was entirely Cargo-controlled, for now. (but paths could be specified as template directories still)

* Could a test be added for a git repository as well?

* We may want to explore `~/.cargo/config` configuration for default templates, I know I'd probably use it!

Also cc @rust-lang/tools, would be good to start hearing thoughts about this as well!

ehiggs commented:
> For example once it's cloned the first time it'll never get updated until it's removed?

There seems to be some confusion here. @gchp's original work cloned the remote repository into `~/.cargo/templates/<repo-name>`. I removed that and merely clone the repo into a tempdir which is thrown away when the resulting Cargo project has been generated. So no local updating is required. Any work like that would be done in `cargo template clone` and/or `cargo template pull`, type commands which aren't in this PR.

So I guess the resulting work should be:

[] git tests using test/git.rs as an example.
[] Options to set the template dir in `~/.cargo/config`
[] Options to set the default template in `~/.cargo/config`

For the default template, I would keep `--lib` and `--bin` use the explicit defaults; but using no flag would use the default template instead of the `--lib` one.

brson commented:
I like the feature and from reading the docs this seems pretty reasonable.

I've always envisioned project templates being part of the platform, so I would expect us to eventually distribute a bunch of them by default, and for IDEs to have options to instantiate them. This implementation doesn't provide for that yet since it only reads from `.cargo` and git repos. That's fine since we can improve the feature in the future, but there is one thing here I'm worried could prevent an important enhancement, and that's the way it gets templates from git repos.

As a step toward including templates in the Rust distribution by default I might expect and encourage somebody to maintain a project collecting templates - a single repo that contains nothing but templates. Thus people would naturally want to instantiate templates out of a subdirectory of a git repo, not from the top-level of the repo. This implementation seems to support one template per repo.

ehiggs commented:
>there is one thing here I'm worried could prevent an important enhancement, and that's the way it gets templates from git repos.

In writing [the example repository](https://github.com/ehiggs/rust-cmdline-skeleton)  I thought that it might be a pain the backside to maintain several repositories for each skeleton. I think the options would be:

- Change the work so it uses a git repo and an optional second argument for a subdirectory. This is straightforward to implement.
- Change the work to be single files and the template just works as a url to the file. This could be a .zip, a .tgz, or maybe even a .patch file which is applied to create the new files and then templated. I would not do something like a Dockerfile or spec since they runs arbitrary scripts and are expected to come from trusted/signed sources. This is more involved but it means that anyone who wants to serve up repos can just toss up a web page with a thin veneer over an object store and serve the templates.

ehiggs commented:
After discussion with @brson on irc, we think this will work:

New lib: `$ cargo new foo --lib`
New bin: `$ cargo new foo --bin`

New local template:

```
$ grep template ~/.cargo/config
template-dir=~/myrusttemplates
default-template=mytemplate
$ cargo new foo --template mytemplate
$ # OR...
$ cargo new foo # no longer assumes `--lib`; uses `default-template setting`.
```

New template based on git repo:
```
$ cargo new foo --template-repo=https://gitlab.com/ehiggs/awesome-template
```

New template based on git repo where the repo is a collection of templates:

```
$ cargo new foo --template-repo=https://gitlab.com/ehiggs/awesome-templates --template=best-template
```

Probably new errors:

```
$ cargo new foo --template=lib # This will run a template called lib; not the builtin `--lib` version.
$ cargo new foo --template=bin # This will run a template called bin; not the builtin `--bin` version.

bors commented:
:umbrella: The latest upstream changes (presumably #3057) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented:
@ehiggs oh thanks for the [previous clarifications](https://github.com/rust-lang/cargo/pull/3004#issuecomment-246953879)! The strategy you and @brson came up with sounds reasonable to me, but I wonder if we could perhaps forgo the `~/.cargo` configuration for now? Or is that required for particular portions?

If you want to rebase I can take another look as well.

ehiggs commented:
I rebased and removed use of the config. So now there is `--template` and `--template-repo`.

The following is supported:

New default library: `cargo new --lib foo` 

New default binary: `cargo new --bin foo`

New templated project based on remote repo:   
`cargo new --template-repo https://rustrepos.org/mytemplate foo`

New templated project selecting a single template from a bunch in the repository:  
`cargo new --template-repo https://rustrepos.org/many-repos --template mytemplate foo`

Local repo (untested ⚠️ ): `cargo new --template-repo file://path/to/template foo`

Local directory: `cargo new --template-repo /path/to/template foo`

Error (template with no associated repo): `cargo new --template mytemplate foo`

TheNeikos commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -85,6 +97,76 @@ struct CargoNewConfig {
+
+    let template_dir = match (opts.template_repo, opts.template) {
+        // given template is a remote git repository & needs to be cloned
+        // This will be cloned to .cargo/templates/<repo_name> where <repo_name>
```
This doesn't seem to be true anymore, I think that's a left over from previous iterations.

bors commented:
:umbrella: The latest upstream changes (presumably #3185) made this pull request unmergeable. Please resolve the merge conflicts.

ehiggs commented:
I think travis had some network trouble:

```
registry-b44aed2c1599b15a: /home/travis/.cargo/registry/src/github.com-88ac128001ac3a9a/libgit2-sys-0.4.5/libgit2/src/global.c:33: git__on_shutdown: Assertion `count <= 8 && count > 0' failed.

error: Process didn't exit successfully: `/home/travis/build/rust-lang/cargo/target/x86_64-unknown-linux-gnu/debug/registry-b44aed2c1599b15a` (signal: 6, SIGABRT: process abort signal)

To learn more, run the command again with --verbose.

make: *** [test-unit-x86_64-unknown-linux-gnu] Error 101
```

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -85,6 +97,87 @@ struct CargoNewConfig {
+            println!("Checking out repo to dir: {}", template_dir.path().display());
+
+            try!(config.shell().status("Cloning", repo_url));
+            match Repository::clone(repo_url, &*template_dir.path()) {
```
Could this use the support in the `sources::git` module? That way it'd handle cloning concerns like proxies and such.

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -85,6 +97,87 @@ struct CargoNewConfig {
+            match Repository::clone(repo_url, &*template_dir.path()) {
+                Ok(_) => {},
+                Err(e) => {
+                    return Err(human(format!( 
```
Could this use `chain_error` and `try!`?

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -85,6 +97,87 @@ struct CargoNewConfig {
+        // given template is a remote git repository & needs to be cloned
+        TemplateType::GitRepo(repo_url) => {
+            let template_dir = try!(TempDir::new(name));
+            println!("Checking out repo to dir: {}", template_dir.path().display());
```
Debug print?

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -366,6 +461,10 @@ fn existing_vcs_repo(path: &Path, cwd: &Path) -> bool {
 }
 
+fn file(p: &Path, contents: &[u8]) -> io::Result<()> {
+        try!(File::create(p)).write_all(contents)
```
extra indent?

alexcrichton replied:
I also think there may be a helper for this in `cargo::util::paths`? If not, could you add one there?

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -412,74 +510,84 @@ fn mk(config: &Config, opts: &MkOptions) -> CargoResult<()> {
-        }
+    // make sure that the template exists
+    if fs::metadata(&template_path).is_err() {
+        return Err(human(format!("Template `{}` not found", template_path.display())))
```
You can use the `bail!` macro here for a bit more ergonomic bailing out, and also stylistically Cargo error messages start with a lowercase letter.

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -412,74 +510,84 @@ fn mk(config: &Config, opts: &MkOptions) -> CargoResult<()> {
+
+    let mut data = BTreeMap::new();
+    data.insert("name".to_owned(), name.to_owned());
+    let author_value = format!("{}", toml::Value::String(author));
```
Is there a way to express a list in handlebars syntax? It seems somewhat odd to me to assume that we always want TOML escaping here?

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -412,74 +510,84 @@ fn mk(config: &Config, opts: &MkOptions) -> CargoResult<()> {
-        if let Some(src_dir) = path_of_source_file.parent() {
-            try!(fs::create_dir_all(src_dir));
+        let mut template_str = String::new();
+        File::open(&entry_path).unwrap().read_to_string(&mut template_str).unwrap();
```
These `unwrap` calls should be translated to errors.

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -412,74 +510,84 @@ fn mk(config: &Config, opts: &MkOptions) -> CargoResult<()> {
+
+        // create the new file & render the template to it
+        let mut dest_file = OpenOptions::new().write(true).create(true).open(&dest_path).unwrap();
+        handlebars.template_renderw(&template_str, &Context::wraps(&data), &mut dest_file).ok();
```
This error shouldn't be ignored, right?

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -412,74 +510,84 @@ fn mk(config: &Config, opts: &MkOptions) -> CargoResult<()> {
-    }
+
+        // create the new file & render the template to it
+        let mut dest_file = OpenOptions::new().write(true).create(true).open(&dest_path).unwrap();
```
This `unwrap` should be an error (I think there may also be a utility function for this)

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -541,16 +649,105 @@ fn global_config(config: &Config) -> CargoResult<CargoNewConfig> {
+fn walk_template_dir(dir: &Path, cb: &mut FnMut(DirEntry)) -> CargoResult<()> {
+    let attr = try!(fs::metadata(&dir));
+
+    let ignore_files = vec!(".gitignore");
```
Typically we use `vec![]` instead of `vec!()`

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -541,16 +649,105 @@ fn global_config(config: &Config) -> CargoResult<CargoNewConfig> {
+    let attr = try!(fs::metadata(&dir));
+
+    let ignore_files = vec!(".gitignore");
+    let ignore_types = vec!("png", "jpg", "gif");
```
This seems like a bit of an odd ignore list, shouldn't we try to walk over everything?

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -541,16 +649,105 @@ fn global_config(config: &Config) -> CargoResult<CargoNewConfig> {
+                    try!(walk_template_dir(&entry.path(), cb));
+                }
+            } else {
+                if let &Some(extension) = &entry.path().extension() {
```
Could these `&` sigils get removed?

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -85,6 +97,87 @@ struct CargoNewConfig {
+            } else {
+                try!(create_lib_template(&temp_templates_path));
+            }
+            TemplateDirectory::Temp(temp_templates_dir)
```
If we are using a builtin template, could we avoid writing out to the filesystem? That shouldn't be strictly required, right?

Perhaps the template could get loaded into memory here in all cases?

ehiggs commented:
I've fixed it so `--bin` and `--lib` no longer use the filesystem to write the template in an intermediate step. I've also pushed a lot of the template functionality to `cargo/util/templates.rs`.

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -59,12 +65,14 @@ impl<'a> NewOptions<'a> {
         let is_lib = if !bin {
             true
-        }
+       }
```
Stray change?

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -85,6 +95,58 @@ struct CargoNewConfig {
+        TemplateType::GitRepo(repo_url) => {
+            let template_dir = try!(TempDir::new(name));
+            try!(config.shell().status("Cloning", repo_url));
+            try!(Repository::clone(repo_url, &*template_dir.path()).chain_error(|| {
```
Could this use the shared infrastructure in `sources::git` for cloning to handle bits like authentication, retries, etc?

ehiggs replied:
I guess the way forward here is to change `GitCheckout::clone_repo` and change source from a Path to a String so it can handle remote urls?

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -412,73 +489,46 @@ fn mk(config: &Config, opts: &MkOptions) -> CargoResult<()> {
-        }
+        // create the new file & render the template to it
+        let mut dest_file = try!(OpenOptions::new().write(true)
+                                 .create(true)
```
I think this can just be `File::create` rather than going through `OpenOptions`, right?

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -541,16 +613,127 @@ fn global_config(config: &Config) -> CargoResult<CargoNewConfig> {
+    let ignore_files = vec![".gitignore"];
+
+    if attr.is_dir() {
+        for entry in try!(fs::read_dir(dir)) {
```
You can cut down on the indentation here with an early return of `if !attr.is_dir()` perhaps?

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -541,16 +613,127 @@ fn global_config(config: &Config) -> CargoResult<CargoNewConfig> {
+            let entry = try!(entry);
+            let attr = try!(fs::metadata(&entry.path()));
+            if attr.is_dir() {
+                if !&entry.path().to_str().unwrap().contains(".git") {
```
these calls to `unwrap()` should be handled gracefully to avoid panicking

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -541,16 +613,127 @@ fn global_config(config: &Config) -> CargoResult<CargoNewConfig> {
+
+fn abs_to_rel_path(abspath: &Path, rel_part: &Path) -> CargoResult<PathBuf> {
+    let abspath_str= try!(abspath.to_str().ok_or(
+            human(format!("Could not convert path to string: {}", abspath.display()))));
```
Stylistically Cargo error messages typically start with lowercase letters (here and a few other locations too)

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -541,16 +613,127 @@ fn global_config(config: &Config) -> CargoResult<CargoNewConfig> {
+    }
+}
+
+fn abs_to_rel_path(abspath: &Path, rel_part: &Path) -> CargoResult<PathBuf> {
```
Could this be implemented with `strip_prefix` from the standard library ?

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -541,16 +613,127 @@ fn global_config(config: &Config) -> CargoResult<CargoNewConfig> {
+// how many there are. Files in the root of the new project direcotory will have two
+// components, anything more than that means the file is in a sub-directory, so we need
+// to create it.
+fn create_deep_subdirs(path: &Path) {
```
I'm not sure I quite understand why this function exists, couldn't this just be an unconditional:

```rust
fs::create_dir_all(&path.parent().unwrap());
```

(modulo error handling and unwrap handling)

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -541,16 +613,127 @@ fn global_config(config: &Config) -> CargoResult<CargoNewConfig> {
+name = "{{name}}"
+version = "0.1.0"
+authors = [{{toml-escape author}}]
+"#)));
```
Oh I think the default template has `[dependencies]` nowadays as well (an empty section)

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -412,73 +489,46 @@ fn mk(config: &Config, opts: &MkOptions) -> CargoResult<()> {
+        let template_str = try!(template.template());
+        //let dest_file_name = try!(abs_to_rel_path(entry_path, template_path));
+        //let mut dest_path = PathBuf::from(path).join(dest_file_name);
+        let dest_path = PathBuf::from(path).join(template.path());
```
I think this can just be `path.join(..)`

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -85,6 +95,58 @@ struct CargoNewConfig {
+fn get_input_template(config: &Config, opts: &MkOptions) -> CargoResult<TemplateSet> {
+    let path = opts.path;
+    let name = opts.name;
+    let templates_dir = config.template_path();
```
I thought we were gonna hold off on the `~/.cargo/templates` for now, right?

alexcrichton commented on src/cargo/ops/cargo_new.rs:
```diff
@@ -114,6 +176,19 @@ fn get_name<'a>(path: &'a Path, opts: &'a NewOptions, config: &Config) -> CargoR
+        (None, Some(_)) => Err(human("A template was given, but no template repository")),
+        (None, None) => Ok(TemplateType::Builtin)
+    }
+}
```
Perhaps this could be moved to the template util module?

alexcrichton commented on src/cargo/util/template.rs:
```diff
@@ -0,0 +1,146 @@
+                          h: &Helper,
+                          _: &Handlebars,
+                          rc: &mut RenderContext) -> Result<(), RenderError> {
+    let param = h.param(0).unwrap();
```
We should likely handle the `unwrap()` here, right?

alexcrichton commented on src/doc/guide.md:
```diff
@@ -408,6 +411,66 @@ will not fail your overall build. Please see the [Travis CI Rust
+[project]
+name = "{{name}}"
+version = "0.1.0"
+authors = [{{{authors}}}]
```
This should use `toml-escape` now, right? Perhaps those various functions could be mentioned here as well?

bors commented:
:umbrella: The latest upstream changes (presumably #3272) made this pull request unmergeable. Please resolve the merge conflicts.

ehiggs commented:
The travis build failures don't appear to be related. Issues with malloc?

brson commented:
I agree the error does not seem related.

It looks like the documentation still reflects the original design, not the updated design. @ehiggs can you update the docs? Maybe include useful examples from [your earlier comment](https://github.com/rust-lang/cargo/pull/3004#issuecomment-252382235).

The final design here has only the `--template` and `--template-repo` flags, and no support for `~/.cargo/template`, as I understand it. So it seems like the focus is on `--template-repo` and creating projects from URLs. You've got to specify a `--template-repo`. If your use case is like @ehiggs original one then you won't specify the (sweeter) `--template` argument.

These are the two cases:

```
cargo new --template-repo https://rustrepos.org/mytemplate foo
cargo new --template-repo https://rustrepos.org/many-repos --template mytemplate foo
```

but it would look sweet to write

```
cargo new --template https://rustrepos.org/mytemplate foo
```

Even with the distinction between a _template repo_ and a _template_ we could make the latter work as a special case. If `--template` is a URL and `--template-repo` isn't specified treat it as if they had written `--template-repo URL`. So I guess having that option makes me feel better about losing the sweeter flag.

Sorry this is taking so long.

cc @alexcrichton just a reminder that this is ongoing

ehiggs commented:
@brson thanks for the feedback. I'll take a look when I next get a window to hack on this.

bors commented:
:umbrella: The latest upstream changes (presumably #3371) made this pull request unmergeable. Please resolve the merge conflicts.

ehiggs commented:
> ```cargo new --template https://rustrepos.org/mytemplate foo```

I like this, and agree that it would be better to have `--template` accept a url. But we also like the idea of using single repositories with multiple project skeletons. Sadly, we can't check out only a single directory from a repository using `--template https://repodomain.org/user/repo/skel-name` but we could try that; if it fails, drop a single directory, try that (`--template ...repo`); and then if it works, use the skel-name dir within the repo. But this kind of guesswork, by making the client make failed requests as part of the design, sucks.

Maybe we can have `--template-repo` become `--template` and `--template` now be known as `--template-subdir` (defaulting to `.`).

bors commented:
:umbrella: The latest upstream changes (presumably #3439) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented:
@ehiggs the idea of `--template` pointing to a URL and `--template-subdir` to assume "maybe it's not at the root of the repo" sounds good to me!

bors commented:
:umbrella: The latest upstream changes (presumably #3504) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented on src/cargo/util/config.rs:
```diff
@@ -65,6 +65,10 @@ impl Config {
 
     pub fn home(&self) -> &Filesystem { &self.home_path }
 
+    pub fn template_path(&self) -> PathBuf {
```
I think this method is unused now, right?

alexcrichton commented on src/cargo/util/template.rs:
```diff
@@ -0,0 +1,231 @@
+pub fn get_template_type<'a>(repo: Option<&'a str>,
+                             subdir: Option<&'a str>) -> CargoResult<TemplateType<'a>> {
+    match (repo, subdir) {
+        (Some(repo_url), _) if repo_url.starts_with("http://")  ||
```
Perhaps we could use `Url` to parse and then dispatch on the scheme if successful?

ehiggs replied:
Sure. It seems there is a bug in the cargo git wrappers, though. `file://` is supported by git as a git protocol. And raw `git2` supports `file://` as a scheme. But the cargo wrappers crap out. Instead of my hack, I think it should probably just be fixed in the wrapper.

alexcrichton replied:
Could you clarify what the bug is? It's intentional that `file://` is a support url scheme for git repos I believe.

ehiggs replied:
In `tests/new.rs`: `git_template` if template treats `file://` as a URL then the following results in an error:
```rust
    assert_that(cargo_process("new").arg("--template").arg(git_project.url().as_str())
                                    .arg("foo")                                 
                                    .env("USER", "foo"),                        
                execs().with_status(0)); 
```

Error:
```
--- stderr
     Cloning file:///Users/ehiggs/src/cargo/target/cit/t1/template1
error: Failed to create project `foo` at `/Users/ehiggs/src/cargo/target/cit/t1/foo`

Caused by:
  Failed to read file for templating: /var/folders/nn/2zxw6dwn17l_vk2xzpj9s_8m0000gp/T/foo.wRUS45BLqSY6/objects/pack/pack-68af5d9cffc2ad3bdb15d6f776a4878c7f8a3670.idx
```
It went away when I treated `file://` as a file.

It's not just the test infrastructure. It also failed when I run it from the cmdline:

```
$ ./target/debug/cargo new --template file://$HOME/src/rust-cmdline-skeleton foo
     Cloning file:///Users/ehiggs/src/rust-cmdline-skeleton
error: Failed to create project `foo` at `/Users/ehiggs/src/cargo/foo`

Caused by:
  Failed to read file for templating: /var/folders/nn/2zxw6dwn17l_vk2xzpj9s_8m0000gp/T/foo.b3gbzHi1Ekl5/objects/pack/pack-47713ed0783c0ed9f74a83830bd6f43b7c8c0146.idx

To learn more, run the command again with --verbose.
```

alexcrichton replied:
Hm could you try passing `--verbose` to see what the underlying git2 error is?

ehiggs replied:
Sure. 

```
Caused by:
  stream did not contain valid UTF-8
```

This could come from [`std::io::append_to_string`](https://github.com/rust-lang/rust/blob/b79081c2eb7fe789f428ffcd20f49549c7ca84ea/src/libstd/io/mod.rs#L329)

alexcrichton replied:
Just to clarify, I think we ended up tracking this down a bare git repository, right?

ehiggs replied:
Yes, the `source::git::fetch` function creates a bare git repository and I needed to do a hard reset to HEAD.

alexcrichton commented on src/doc/guide.md:
```diff
@@ -408,6 +413,89 @@ will not fail your overall build. Please see the [Travis CI Rust
 
+# Templates
+
+Cargo uses the handlebars library to compile the templates used to scaffold
```
Could this hyperlink to the handlebars library?

alexcrichton commented:
Awesome! Just a few minor nits but otherwise r=me

ehiggs commented:
Bah, I parse template strings as URLs and check if are indeed URLs supported by git. If they are URLs but not using a scheme we support, we yield an error. However, apparently `C:\` is parsed as a url...

```
---- simple_template stdout ----
	running `C:\projects\cargo\target\i686-pc-windows-msvc\release\cargo new --template-subdir testtemplate --template C:\projects\cargo\target\i686-pc-windows-msvc\cit\t23\home/.cargo/templates/ foo`
thread 'simple_template' panicked at '
Expected: execs
    but: exited with exit code: 101
--- stdout
--- stderr
error: Failed to create project `foo` at `C:\projects\cargo\target\i686-pc-windows-msvc\cit\t23\foo`
Caused by:
  A URL was given with an unsupported scheme: c
', C:\Users\appveyor\.cargo\registry\src\github.com-1ecc6299db9ec823\hamcrest-0.1.1\src\core.rs:31
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

e.g. the following will print `c`
 ```rust
extern crate url;
use url::Url;

fn main() {
    let u = Url::parse("c:\\hello").unwrap();
    println!("{}", u.scheme());
}
```

alexcrichton commented:
Oh so I figured we would:

* Attempt to parse the template as a URL
* If successful and not `file://`, download a git repo
* If successful and `file://`, or parsing failed, interpret as a path and interpret that as a template

Does that make sense?

ehiggs commented:
Sure. I have a failing test + fix. We don't even need to treat the `file://` as a special case of local dir. We can still treat it as a GitRepo.

I thought I would also add support for ssh scp style syntax (`user@domain.com:path/to/repo`) but I get errors which are basically the same as #3014. Should I leave the code in that handles the scp style identifier with the requisite error; or should we give an error message explaining that it's not supported at the moment.

alexcrichton commented:
Everything looks superb to me! @ehiggs want to squash things down a bit in preparation for merging?

Also if it's ok with you I think I might hold off on merging for a week or so to get this into Rust 1.17 instead of 1.16. This is a relatively major new feature to Cargo so I'd just want to be sure to get ample time in testing and train riding before we ship it on stable. The beta branches next week so it won't take too long!

ehiggs commented:
Great news! Thanks for your patience with this.  I'll look into squashing it.

One last feature/issue: I see that [cookiecutter](https://github.com/audreyr/cookiecutter) also parses filenames and directories as templates. [Here is an example](https://github.com/wdm0006/cookiecutter-pipproject). It's important in python since the equivalent of the `src` directory is almost always the name of the package. So maybe it's not important for Cargo.

alexcrichton commented:
Oh interesting! I'd probably advocate avoiding that for now in Cargo (I don't think there's a *huge* amount of precedent) but we can always add it backwards-compatibly.

ehiggs commented:
Hi,
I've squashed it now and added a further commit which upgrades handlebars to 0.25.

The unsquashed commits are here: https://github.com/ehiggs/cargo/tree/templating

alexcrichton commented:
Ok! I'll r+ some time this coming week once we've branched.

bors commented:
:umbrella: The latest upstream changes (presumably #3618) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented:
Beta was branched let's merge!

@bors: r+

bors commented:
:pushpin: Commit 265452d has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 265452d49d1a004fb732b624f1f1e8496f75de7a -->

bors commented:
:hourglass: Testing commit 265452d with merge 5ef54e0...

ehiggs commented:
Travis has had problems with the OS X queue. Hence the ❌

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/cargo/builds/197170931)

alexcrichton commented:
@bors: retry

bors commented:
:hourglass: Testing commit 265452d with merge f7fb965...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang-libs/cargo/build/1.0.1054), [status-travis](https://travis-ci.org/rust-lang/cargo/builds/197464681)
Approved by: alexcrichton
Pushing f7fb965fd22601e16c8547aae9567179c4eb736d to master...

alexcrichton commented:
🎉 

Thanks so much again @ehiggs for your patience!

ehiggs commented:
Thank you for the patience. Now time to go bash out some templates. 😀

gchp commented:
@ehiggs I think it's safe to say you far out-did my original work :smile: great job!

WiSaGaN commented:
A note that this pull-request has been reverted in https://github.com/rust-lang/cargo/pull/3878 in waiting for a RFC.

ehiggs commented:
Indeed, ongoing discussion is [here](https://internals.rust-lang.org/t/pre-rfc-cargo-templates/5056). It seems there is no consensus yet.
